pkg/dataplane/nginx.go

```diff
@@ -0,0 +1,284 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dataplane
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sync"
+	"syscall"
+	"time"
+
+	"k8s.io/client-go/util/flowcontrol"
+	"k8s.io/ingress-nginx/internal/ingress/controller/process"
+	ngx_template "k8s.io/ingress-nginx/internal/ingress/controller/template"
+	"k8s.io/ingress-nginx/internal/ingress/metric"
+	ing_net "k8s.io/ingress-nginx/internal/net"
+	"k8s.io/ingress-nginx/internal/net/dns"
+	"k8s.io/ingress-nginx/internal/nginx"
+	"k8s.io/ingress-nginx/internal/task"
+	"k8s.io/ingress-nginx/pkg/apis/ingress"
+	"k8s.io/ingress-nginx/pkg/tcpproxy"
+	"k8s.io/ingress-nginx/pkg/util/file"
+	"k8s.io/klog/v2"
+)
+
+// NewNGINXController creates a new NGINX Ingress controller.
+func NewNGINXConfigurer(config *Configuration, mc metric.Collector) *NGINXConfigurer {
+
+	h, err := dns.GetSystemNameServers()
+	if err != nil {
+		klog.Warningf("Error reading system nameservers: %v", err)
+	}
+
+	n := &NGINXConfigurer{
+		isIPV6Enabled:   ing_net.IsIPv6Enabled(),
+		resolver:        h,
+		cfg:             config,
+		syncRateLimiter: flowcontrol.NewTokenBucketRateLimiter(config.SyncRateLimit, 1),
+		stopCh:          make(chan struct{}),
+		ngxErrCh:        make(chan error),
+		configureLock:   &sync.Mutex{},
+		stopLock:        &sync.Mutex{},
+		// TOOD: Right now we will receive the full configuration, but we may want to receive and validate just checksums
+		runningConfig: new(ingress.Configuration),
+		Proxy:         &tcpproxy.TCPProxy{},
+
+		metricCollector: mc,
+
+		command: NewNginxCommand(),
+	}
+
+	// TODO: Change to trigger the full reconciliation
+	getFullReconciliation := func(obj interface{}) error {
+		return nil
+	}
+	n.syncQueue = task.NewTaskQueue(getFullReconciliation)
+
+	onTemplateChange := func() {
+		template, err := ngx_template.NewTemplate(nginx.TemplatePath)
+		if err != nil {
+			// this error is different from the rest because it must be clear why nginx is not working
+			klog.ErrorS(err, "Error loading new template")
+			return
+		}
+
+		n.t = template
+		klog.InfoS("New NGINX configuration template loaded")
+		// We trigger an event for a full reconciliation if template changes
+		n.syncQueue.EnqueueTask(task.GetDummyObject("template-change"))
+	}
+
+	ngxTpl, err := ngx_template.NewTemplate(nginx.TemplatePath)
+	if err != nil {
+		klog.Fatalf("Invalid NGINX configuration template: %v", err)
+	}
+
+	n.t = ngxTpl
+
+	_, err = file.NewFileWatcher(nginx.TemplatePath, onTemplateChange)
+	if err != nil {
+		klog.Fatalf("Error creating file watcher for %v: %v", nginx.TemplatePath, err)
+	}
+
+	filesToWatch := []string{}
+	err = filepath.Walk("/etc/nginx/geoip/", func(path string, info os.FileInfo, err error) error {
```

Comment:
Will it be considered to allow configuration? (I haven't thought of a specific scenario yet, but it may be some extensions, or user-defined behavior